File: src/main/java/net/onelitefeather/stardust/user/UserProperty.java
package net.onelitefeather.stardust.user;

import jakarta.persistence.*;

import java.util.Objects;

@Entity
@Table
public class UserProperty {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private final Long id;

    @Column
    private final String name;

    @Column
    private String value;

    @Column
    private final Byte type;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    public UserProperty(Long id, String name, String value, Byte type, User user) {
        this.id = id;
        this.name = name;
        this.value = value;
        this.type = type;
        this.user = user;
    }

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public Byte getType() {
        return type;
    }

    public User getUser() {
        return user;
    }

    public UserProperty withValue(String value) {
        this.value = value;
        return this;
    }

    public UserProperty withUser(User user) {
        this.user = user;
        return this;
    }

    public UserPropertyType propertyType() {
        return UserPropertyType.getByName(this.name.toUpperCase());
    }

    @SuppressWarnings("unchecked")
    public <T> T getValue() {
        return switch (type) {
            case 1 -> (T) Boolean.valueOf(value);
            case 2 -> (T) Integer.valueOf(value);
            case 3 -> (T) Long.valueOf(value);
            case 4 -> (T) Double.valueOf(value);
            case 5 -> (T) String.valueOf(value);
            default -> (T) value;
        };
    }

    public static UserProperty of(UserPropertyType type) {
        return new UserProperty(
                null,
                type.getName(),
                type.getDefaultValue().toString(),
                type.getType(),
                null
        );
    }

    @Override
    public final boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof UserProperty that)) return false;

        return Objects.equals(id, that.id) && Objects.equals(name, that.name) && Objects.equals(value, that.value) && Objects.equals(type, that.type);
    }

    @Override
    public int hashCode() {
        int result = Objects.hashCode(id);
        result = 31 * result + Objects.hashCode(name);
        result = 31 * result + Objects.hashCode(value);
        result = 31 * result + Objects.hashCode(type);
        return result;
    }

    @Override
    public String toString() {
        return "UserProperty{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", value='" + value + '\'' +
                ", type=" + type +
                '}';
    }
}

File: src/main/kotlin/net/onelitefeather/stardust/user/UserProperty.kt
package net.onelitefeather.stardust.user

import jakarta.persistence.*

@Entity
@Table
data class UserProperty(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long?,

    @Column
    val name: String = "",

    @Column
    val value: String = "",

    @Column
    val type: Byte = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    val user: User? = null) {

    constructor() : this(null)

    fun <T : Any> getValue(): T? {
        val result = when (this.type.toInt()) {
            0 -> this.value
            1 -> this.value.toIntOrNull()
            2 -> this.value.toBooleanStrictOrNull()
            3 -> this.value.toDoubleOrNull()
            4 -> this.value.toFloatOrNull()
            5 -> this.value.toShortOrNull()
            6 -> this.value.toByteOrNull()
            else -> null
        }
        return result as T?
    }

    @Override
    override fun toString(): String {
        return this::class.simpleName + "(id = $id , key = $name , value = $value )"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is UserProperty) return false

        if (id != other.id) return false
        if (name != other.name) return false
        if (value != other.value) return false
        if (type != other.type) return false
        if (user != other.user) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id?.hashCode() ?: 0
        result = 31 * result + name.hashCode()
        result = 31 * result + value.hashCode()
        result = 31 * result + type
        result = 31 * result + (user?.hashCode() ?: 0)
        return result
    }
}